java/arcs/core/storage/driver/Database.kt

```diff
@@ -69,36 +80,72 @@ data class DatabaseStorageKey(
                 "Not a valid DatabaseStorageKey: $rawKeyString"
             }
 
-            val persistent = match.groupValues[2] == "persistent"
-            return DatabaseStorageKey(match.groupValues[3], persistent, match.groupValues[1])
+            val entitySchemaHash = match.groupValues[1]
+            val dbName = match.groupValues[2]
+            val persistent = match.groupValues[3] == "persistent"
+            val unique = match.groupValues[4]
+            return DatabaseStorageKey(unique, entitySchemaHash, persistent, dbName)
         }
     }
 }
 
 /** [DriverProvider] which provides a [DatabaseDriver]. */
-class DatabaseDriverProvider : DriverProvider {
+class DatabaseDriverProvider(
+    /**
+     * Function which will be used to determine, at runtime, which [Schema] to associate with its
+     * hash value embedded in a [DatabaseStorageKey].
+     */
+    private val schemaLookup: (String) -> Schema?
+) : DriverProvider {
     init {
         DriverFactory.register(this)
     }
 
     override fun willSupport(storageKey: StorageKey): Boolean =
-        storageKey is DatabaseStorageKey
+        storageKey is DatabaseStorageKey && schemaLookup(storageKey.entitySchemaHash) != null
 
     override suspend fun <Data : Any> getDriver(
         storageKey: StorageKey,
         existenceCriteria: ExistenceCriteria
     ): Driver<Data> {
-        require(storageKey is DatabaseStorageKey) {
+        val databaseKey = requireNotNull(storageKey as? DatabaseStorageKey) {
             "Unsupported StorageKey: $storageKey for DatabaseDriverProvider"
         }
-        return DatabaseDriver(storageKey, existenceCriteria)
+        requireNotNull(schemaLookup(databaseKey.entitySchemaHash)) {
+            "Unsupported DatabaseStorageKey: No Schema found with hash: " +
+                databaseKey.entitySchemaHash
+        }
+        return DatabaseDriver(storageKey, existenceCriteria, schemaLookup)
+    }
+
+    // Both equals(other) and hashCode() should treat all instances of DatabaseDriverProvider as
+    // equivalent, so that the DriverFactory keeps one instance, even if the class is instantiated
+    // more than once.
```

Comment:
Would it be possible to declare this class as `object` instead? IIUC that would make it a singleton and you wouldn't be able to instantiate more than one.

Reply:
Good point, that would probably work well... will do.